elitzur-avro/src/main/scala/com/spotify/elitzur/converters/avro/dynamic/DynamicAccessorCompanion.scala

```diff
@@ -0,0 +1,91 @@
+import com.spotify.elitzur.converters.avro.AvroElitzurConversionUtils.byteBufferToByteArray
+import com.spotify.elitzur.validators.{
+  BaseCompanion,
+  BaseValidationType,
+  SimpleCompanionImplicit,
+  Validator
+}
+
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe.{TypeTag, typeOf}
+import java.{util => ju}
+import collection.JavaConverters._
+
+//scalastyle:off line.size.limit structural.type
+class DynamicAccessorCompanion[T: TypeTag, LT <: BaseValidationType[T]: ClassTag: ({type L[x] = SimpleCompanionImplicit[T, x]})#L] extends Serializable {
+//scalastyle:on line.size.limit structural.type
+
+  private val companion: BaseCompanion[T, LT] =
+    implicitly[SimpleCompanionImplicit[T, LT]].companion
+  private[dynamic] val validationType: String = companion.validationType
+  private def parseUnsafe(data: Any): Any = companion.parse(data.asInstanceOf[T])
+
+  private def parseAvro(o: Any): Any =
+    typeOf[T] match {
+      // String in Avro can be stored as org.apache.avro.util.Utf8 (a subclass of Charsequence)
+      // which cannot be cast to String as-is. The toString method is added to ensure casting.
+      case t if t =:= typeOf[String] => parseUnsafe(o.toString)
+      // ByteBuffer in Avro to be converted into Array[Byte] which is the the format that Validation
+      // type expects the input the input to be in.
+      case t if t =:= typeOf[Array[Byte]] => parseUnsafe(
+        byteBufferToByteArray(o.asInstanceOf[java.nio.ByteBuffer]))
+      case _ => parseUnsafe(o)
+    }
+
+  // TODO: Optimize below method by introducing changes to Elitzur-Core to allow non-implicit driven
+  // wiring of Validators
+  //scalastyle:off line.size.limit
+  private[dynamic] def getValidator(modifiers: List[ValidatorOp])(implicit m: MetricsReporter): Validator[Any] = {
+  //scalastyle:on line.size.limit
+    modifiers match {
+      case Nil => implicitly[Validator[LT]].asInstanceOf[Validator[Any]]
+      case OptionValidatorOp :: Nil => implicitly[Validator[Option[LT]]]
+        .asInstanceOf[Validator[Any]]
+      case ArrayValidatorOp :: Nil => implicitly[Validator[Seq[LT]]]
+        .asInstanceOf[Validator[Any]]
+      case ArrayValidatorOp :: OptionValidatorOp :: Nil => implicitly[Validator[Seq[Option[LT]]]]
+        .asInstanceOf[Validator[Any]]
+      case OptionValidatorOp :: ArrayValidatorOp :: Nil => implicitly[Validator[Option[Seq[LT]]]]
+        .asInstanceOf[Validator[Any]]
+      case OptionValidatorOp :: ArrayValidatorOp :: OptionValidatorOp :: Nil =>
+        implicitly[Validator[Option[Seq[Option[LT]]]]].asInstanceOf[Validator[Any]]
+      case _ => throw new Exception(s"Unsupported validator operation: ${modifiers.mkString(",")}")
+    }
+  }
+
+  private[dynamic] def getPreprocessorForValidator(modifiers: List[ValidatorOp]): Any => Any = {
+    val baseFn: Any => Any = (v: Any) => parseAvro(v)
+    modifiers.reverse.foldLeft(baseFn)((a, c) => (v: Any) => c.preprocessorOp(v, a))
```

Comment:
So this is essentially composing the extract/preprocessing code for each level of nesting if Im understanding correctly? can you add a comment just explaining that? feel like it's a bit tricky to understand since there's no type safety